test/DateYearDurationTest.cpp

```diff
@@ -0,0 +1,977 @@
+               DurationParseException);
+  ASSERT_THROW(DayTimeDuration::parseXsdDayTimeDuration("P0DABH0M0S"),
+               DurationParseException);
+}
+
+//______________________________________________________________________________
+TEST(DayTimeDuration, checkToAndFromBits) {
+  auto d1 = DayTimeDuration(DayTimeDuration::Type::Positive, 1, 23, 23, 59.99);
+  auto bits = d1.toBits();
+  d1 = DayTimeDuration::fromBits(bits);
+  DayTimeDuration::DurationValue dv = d1.getValues();
+  ASSERT_EQ(dv.days_, 1);
+  ASSERT_EQ(dv.hours_, 23);
+  ASSERT_EQ(dv.minutes_, 23);
+  EXPECT_NEAR(dv.seconds_, 59.99, 0.001);
+
+  auto d2 =
+      DayTimeDuration(DayTimeDuration::Type::Negative, 1048574, 3, 0, 0.99);
+  bits = d2.toBits();
+  d2 = DayTimeDuration::fromBits(bits);
+  dv = d2.getValues();
+  ASSERT_EQ(dv.days_, 1048574);
+  ASSERT_EQ(dv.hours_, 3);
+  ASSERT_EQ(dv.minutes_, 0);
+  EXPECT_NEAR(dv.seconds_, 0.99, 0.001);
+}
+
+//______________________________________________________________________________
+TEST(DayTimeDuration, DurationOverflowException) {
+  try {
+    [[maybe_unused]] auto d1 =
+        DayTimeDuration(DayTimeDuration::Type::Positive, 643917423, 4, 7, 1);
+    FAIL() << "DurationOverflowException was expected.";
+  } catch (const DurationOverflowException& e) {
+    EXPECT_STREQ(e.what(),
+                 "Overflow exception raised by DayTimeDuration, please provide "
+                 "smaller values for xsd:dayTimeDuration.");
+  }
+}
+
+//______________________________________________________________________________
+DayTimeDuration::DurationValue toAndFromMilliseconds(int days, int hours,
+                                                     int minutes,
+                                                     double seconds) {
+  // to milliseconds
+  auto millisecDays = static_cast<long long>(days) * 24 * 60 * 60 * 1000;
+  auto millisecHours = static_cast<long long>(hours) * 60 * 60 * 1000;
+  auto millisecMinutes = static_cast<long long>(minutes) * 60 * 1000;
+  auto millisecSeconds = static_cast<long long>(std::round(seconds * 1000));
+  auto totalMilliseconds =
+      millisecDays + millisecHours + millisecMinutes + millisecSeconds;
+
+  // from milliseconds
+  long long remainingMilliseconds = totalMilliseconds;
+  days = remainingMilliseconds / (24 * 60 * 60 * 1000);
+  remainingMilliseconds %= 24 * 60 * 60 * 1000;
+  hours = remainingMilliseconds / (60 * 60 * 1000);
+  remainingMilliseconds %= 60 * 60 * 1000;
+  minutes = remainingMilliseconds / (60 * 1000);
+  remainingMilliseconds %= 60 * 1000;
+  seconds = remainingMilliseconds / 1000.0;
+  return {days, hours, minutes, seconds};
+}
+
+//______________________________________________________________________________
+TEST(DayTimeDuration, checkInternalConversionForLargeValues) {
+  int maxDays = static_cast<int>(DayTimeDuration::maxDays);
+  ad_utility::SlowRandomIntGenerator randomDay{1000000, maxDays};
```

Comment:
Should the upper limit also be larger than the maxDays to catch the out of bounds here?